### ISSUE TITLE
Dealing with settings when cross-building is quite tricky

### ISSUE THREAD

**Issue by theodelrieu** [conan-io/conan#2098]
Hi!

Lately I've been doing cross-building, on Android mostly. The main issue I faced is that `self.settings` are the **target** settings.

That's a bit confusing when one tries to restrict some of those settings (e.g. the `os` for a specific toolchain package). I either use `platform.system()` directly, or `tools.detected_architecture()` for the `arch`.

Do you think it would be possible to add something like `self.host_settings` when `tools.cross_building() == True`? Having an easy way to access those settings would be really helpful.

**Comment by theodelrieu** [comment]
> ... only `os`, and `arch` makes sense

Wouldn't `compiler` be useful too? I don't use it personally though, but if someone does not want to use a toolchain recipe, they might want to set the cross compiler themselves.

>But maybe `host_settings` are a bit confusing

Agreed, I even confused the terms `host`, `target`, and `build`...
I think following the [GNU convention](https://www.gnu.org/software/automake/manual/html_node/Cross_002dCompilation.html) would be a good start.

`self.settings` are actually the `host` settings (I'm not proposing to change the name, that would break everything).

About the UI I imagined something like the following:

```python
    build_compiler = self.cross_settings["build"].compiler
    host_os = self.cross_settings["host"].os # equivalent to self.settings.os
    target_arch = self.cross_settings["target"].arch
```

`cross_settings` would always be populated, the default being `build` == `host` == `target`.

**Comment by memsharded** [comment]
> Wouldn't compiler be useful too? 

But compiler should be in ``self.settings``, I don't see 2 compilers here

``self.settings`` are the **target** settings. In the early days, they were always the same as the host settings. I think we probably want to use "host" and "target" naming. When I said about ``host_settings``, is because maybe the easiest would be something like ``host_os`` and ``host_arch``. I have to review ``OSInfo``, probably that is the best way to "standardize" it.

**Comment by theodelrieu** [comment]
>`self.settings` are the **target** settings. In the early days, they were always the same as the host settings.

Referring to the GNU `automake` page that I've mentioned above:

> The --host and --build options are usually all we need for cross-compiling. The only exception is if the package being built is itself a cross-compiler: we need a third option to specify its target architecture.
>
>**--target=target**
>
>    When building compiler tools: the system for which the tools will create output. 

In this specific case I don't think `self.settings` represent the **target** settings, since the conan package will contain a cross compiler, not the output they produce.

> But `compiler` should be in `self.settings`, I don't see 2 compilers here

Indeed, I was wrong, `self.settings.compiler` is the compiler with which the package was built.
In that case I agree that having `self.cross_settings["build"].compiler` makes no sense.

> maybe the easiest would be something like `host_os` and `host_arch`

Yes, and having access to `build_os/arch` and `target_os/arch` would be useful as well.

**Comment by memsharded** [comment]
Oh, yes, now I see better what you mean, and I understand the confusion. When talking about packages containing cross-compilers, the package settings are the target settings, but no model for the host settings. We have to think about it, indeed.

**Comment by memsharded** [comment]
This package: https://github.com/lasote/conan-android-ndk/blob/master/conanfile.py
Uses options to control it. I think this is an indication that something can be improved here.

**Comment by lasote** [comment]
Yes, thanks for your feedback with this @theodelrieu, very interesting.

**Comment by theodelrieu** [comment]
There's also an edge-case with cross compilers, for example the Android NDK's clang can target multiple archs with a single binary.

**Comment by sixten-hilborn** [comment]
Hi, I've had similar thoughts too with having some kind of host settings.
My concrete case was to wrap Emscripten in a package and use it via `build_requires`. Some of my problems were that the `os` setting isn't useful as it's always set to `Emscripten`, and that the `CMake` build helper would not work correctly inside the emscripten package. With host settings I'd have something like this:

    os: Emscripten
    compiler: emcc
    arch: [asmjs, wasm]
    host_os: [Windows, Linux, ...]
    host_compiler: [Visual Studio, gcc, ...]
    host_arch: [x86, ...]

For packages/toolchains that support multiple archs I think using [package_id](http://docs.conan.io/en/latest/howtos/define_abi_compatibility.html?highlight=package_id#defining-a-custom-package-id) and checking the arch in `configure()` is the solution.

**Comment by memsharded** [comment]
> There's also an edge-case with cross compilers, for example the Android NDK's clang can target multiple archs with a single binary.

This case can be managed with ``settings`` (for example, removing the ``arch`` setting from the declaration in the package, means that it will target all architectures), or together with the ``build, package`` methods, with the ``package_id()`` you can define which packages target which architectures.

**Comment by memsharded** [comment]
There might be 3 approaches, from what I have checked so far:

- Adding a ``host_settings`` attribute to conanfile. This is the one that requires more implementation, adding the model, adding the user interface, plugin the model into the package_id computation, etc
Also, auto-detection

- Adding some ``host_os``, ``host_arch`` new settings to default settings.yml, with None as valid values. This requires very little code to add, practically only settings.yml, maybe some auto-detection.

- As these packages are not very numerous, maybe propose some option-based approach. A bit of a workaround but could be enough:

```python
class Pkg(ConanFile):
     settings = ... # the same
     options = {"host_os": ["Win", "Linux"], # my cross toolchain only exists for these
                      "host_arch": ["x86", "x86_64"]} 
     def configure(self):
         self.options.host_os = tools.detect_os()
         self.options.host_arch = tools.detect_archictecture()
```
This model might be good for most cases, will manage to have different binaries IDs for different platforms. If the auto-detection doesn't work, it could be not-defined, and force the user to define. Not a big deal, the error message will be clear, and defining it in command line or in profile is easy.

**Comment by theodelrieu** [comment]
Just to avoid further confusion: 
If we're following the GNU convention,  **host** is the package consumer system, not the package creator's. The latter is named  **build**, whereas the **target** is only relevant when dealing with cross-compilers.

Conan packages' `self.settings` are the **host** settings.

I believe we should be very careful to the wording here, otherwise this will mislead a lot of users (e.g. when cross-building an Autotools-based project, I know it did mislead me the first time).

>As these packages are not very numerous, maybe propose some option-based approach.

This option seems a bit hacky, I'd rather have first class support for cross-built packages, especially for mobile devices. 

>Adding some `host_os`, `host_arch` new settings to default `settings.yml`, with None as valid values.

I don't think `None` should be allowed. By default **build**, **host** and **target** should be equivalent.
I do like this approach, only 4 settings would be added:

* `build_os`
* `build_arch`
* `target_os`
* `target_arch`

The first two would be auto-detected by default.

One last thought, I don't think the `build_*` settings should have an impact on the package ID.

**Comment by memsharded** [comment]
> Conan packages' ``self.settings`` are the host settings.

Oh yes, regarding GNU convention this is right. I was actually pretty confused myself, because out of that convention, I feel many users won't understand "host" as the package consumer, but as the builder. 

So, lets say  I am in a Linux box, compiling with gcc for Android:
- build = Linux
- host = Android
- target = ? Could you please tell me an example of this? I understand that I am already dealing with a cross-compiler, am I?

Maybe we should run a survey to get more feedback about the naming. Though better clarify the core functionality.

> I don't think None should be allowed. By default build, host and target should be equivalent.

The problem when adding settings with a default value, they change packageIDs. Doing this will automatically invalidate all existing binary packages in all conan installations, including conan-center, but also in all private repos. Massive breakage. Using the convention ``=None`` meaning = it is the same as the host one, doesn't change the packageID.

**Comment by sixten-hilborn** [comment]
IMO the GNU names are unintuitive. They're described here: https://www.gnu.org/software/automake/manual/html_node/Cross_002dCompilation.html

--target is used for cross-compiling a cross-compiler, and specifies arch etc. for the cross-compiler output.

**Comment by theodelrieu** [comment]
@sixten-hilborn 

>IMO the GNU names are unintuitive.
I agree that `--target` and `--host` seem quite similar and thus confusing. On the other hand there's a lot of libraries based on Autotools, I'm not sure that adding a different meaning would be helpful (especially if we use the same words!).

>--target is used for cross-compiling a cross-compiler

You have to use `--target` when compiling a cross-compiler, even if you do not cross-compile it.


@memsharded 

In your case the `target` option is not useful, here is a case where it is:

Building a GCC on Linux, to run on Linux, to produce output for armv7

* build = Linux, x86_64
* host = Linux, x86_64
* target = Android, armv7


>Maybe we should run a survey to get more feedback about the naming. Though better clarify the core functionality.

Good idea.

>Using the convention =None meaning = it is the same as the host one, doesn't change the packageID.

I guess we could apply the same thing when `build` and `target` are the same than `host`, but that might complicate the code.
It's more of a style choice here, I'd be happy with both options.

**Comment by SSE4** [comment]
that applies not only to the cross-compilers, but for all build tools in general (e.g. 7z, nasm, yasm, ninja, perl, cmake, etc).
also, many people are cross-compiling Linux->Windows and vice versa. in such case, building ffmpeg for Windows on Linux shouldn't invoke installation of Windows version of yasm (which will not run anyway).
Android - same story, cross-compiling Windows->Android shouldn't invoke installation of Android (e.g. ARM) yasm binary (which may be even non-existing).

**Comment by theodelrieu** [comment]
I've been giving some thought to this issue lately, maybe following the GNU convention is not the best idea.

For example, LLVM/Clang can target a lot of architectures with a single binary (a fat compiler if you will).

I don't think we could have a nice descriptive package of such a compiler by following the GNU convention.

I still think handling the cross-compiler case is important to have a great model, but that could be achieved incrementally. I don't want 99% of use cases crippled because of this one in particular.

**Comment by lasote** [comment]
> For example, LLVM/Clang can target a lot of architectures with a single binary (a fat compiler if you will).

In that case, target.os/target.arch can be declared, but the `package_id()` could do:

```
def package_id():
    self.info.target.os = "ANY"
    self.info.taget.arch = "ANY"
```

So the same LLVM/Clang compiler will be retrieved. Did I get it right?

**Comment by theodelrieu** [comment]
Oh right, I haven't thought about that :)

**Comment by memsharded** [comment]
This has been released in conan 1.0.0-beta, please have a look to http://docs.conan.io/en/latest/conan1.0.html and report as soon as possible, so things can be fixed before the final release. Thanks very much!
